File: run-tests.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Basename;

sub red {
    return "\x1B[31m" . $_[0] . "\x1b[0m"
}

sub green {
    return "\x1B[32m" . $_[0] . "\x1b[0m"
}

my @test_files = glob('./tests/auto/*.u');
my $padding = 0;

foreach my $test_file (@test_files) {
    chomp($test_file);
    my $size = length(basename($test_file));
    if ($size > $padding) {
        $padding = $size;
    }
}
$padding += 3;

my $flags = '--eliminate-tail-recursion --lib-dir lib';

for my $test_file (@test_files) {
    my $output;
    my $compiler_output = `./compiler $flags $test_file 2>&1`;
    
    printf("%-${padding}s", basename($test_file));

    if (length($compiler_output) > 0) {
        $output = red('Compilation Failed');
    } else {
        my $test_output = `./output`;
        chomp($test_output);
        
        if ($test_output eq 'PASS') {
            $output = green('PASS');
        } elsif ($test_output eq 'FAIL') {
            $output = red('FAIL');
        } elsif ($test_output eq '') {
            $output = red('No output');
        } else {
            $output = red('Malformed output: ' . $test_output);
        }
    }
    
    print($output . "\n");
}

File: run-tests.py
#!/usr/bin/python
from subprocess import Popen, PIPE, STDOUT
import os
import glob
import sys

def red(text):
    return '\x1B[31m' + text + '\x1B[0m'

def green(text):
    return '\x1B[32m' + text + '\x1B[0m'

test_files = []
if len(sys.argv) > 1:
    sys.argv.pop(0)
    for arg in sys.argv:
        test_files.append('./tests/auto/' + arg)
else:
    test_files = glob.glob('./tests/auto/*.u')

padding = 0
for test_file in test_files:
    padding = max(padding, len(os.path.basename(test_file))+1)

compile_cmd = [
    './compiler',
    '--eliminate-tail-recursion',
    '--lib-dir', 'lib',
    ''
]

for test_file in test_files:
    if not os.path.exists(test_file):
        print(red('File does not exist: ' + test_file))
        continue

    compile_cmd[-1] = test_file
    compile_output = Popen(compile_cmd, stdout=PIPE, stderr=STDOUT).stdout.read()
    output = ''

    if len(compile_output) > 0:
        output = red('Compilation Failed')
    else:
        p = Popen('./output', stdout=PIPE, stderr=STDOUT)
        test_output = p.stdout.read()
        test_output = test_output.decode('ASCII').strip()
        ret = p.poll()

        if ret < 0:
            output = red('Execution Failed')
        elif test_output == 'PASS':
            output = green('PASS')
        elif test_output == 'FAIL':
            output = red('FAIL')
        elif test_output == '':
            output = red('No Output')
        else:
            output = red('Bad Output')

    print(os.path.basename(test_file).ljust(padding) + output)
